app/src/main/java/co/runloop/influencer/fragment/ContactsFragment.java

```diff
@@ -0,0 +1,110 @@
+import android.support.v7.widget.LinearLayoutManager;
+import android.support.v7.widget.RecyclerView;
+import android.util.Log;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+
+import java.util.List;
+
+import co.runloop.influencer.R;
+import co.runloop.influencer.adapter.ContactsAdapter;
+import co.runloop.influencer.data.ContactsProvider;
+import co.runloop.influencer.model.Contact;
+import co.runloop.influencer.viewmodel.ContactsViewModel;
+
+public class ContactsFragment extends BaseFragment {
+
+    private static final String TAG = ContactsFragment.class.getSimpleName();
+
+    private static final int READ_CONTACT_PERM_RC = 101;
+
+    public static ContactsFragment newInstance() {
+        Bundle args = new Bundle();
+        ContactsFragment fragment = new ContactsFragment();
+        fragment.setArguments(args);
+
+        return fragment;
+    }
+
+    private RecyclerView recyclerView;
+    private ContactsAdapter adapter;
+    private ContactsViewModel contactsViewModel;
+
+    @Override
+    public void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        adapter = new ContactsAdapter();
+        if (isActionBarAvailable()) {
+            getBaseActivity().getSupportActionBar().setTitle(R.string.contacts);
+        }
+    }
+
+    @Nullable
+    @Override
+    public View onCreateView(@NonNull LayoutInflater inflater,
+                             @Nullable ViewGroup container,
+                             @Nullable Bundle savedInstanceState) {
+        View root = inflater.inflate(R.layout.frag_contacts_list, container, false);
+
+        recyclerView = root.findViewById(R.id.frag_contacts_list_recyclerview);
+        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
+        recyclerView.setAdapter(adapter);
+
+        contactsViewModel = ViewModelProviders
+                .of(this)
+                .get(ContactsViewModel.class);
+        contactsViewModel.getContacts().observe(this, (@Nullable List<Contact> contacts) -> {
+            adapter.setContacts(contacts);
+            adapter.notifyDataSetChanged();
+        });
+        if (contactsIsAvailable()) {
+            contactsViewModel.onPermissionGranted();
+            if(contactsViewModel.getContacts().getValue() == null) {
+                contactsViewModel.loadAll();
+            }
+        }
+
+        return root;
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode,
+                                           @NonNull String[] permissions,
+                                           @NonNull int[] grantResults) {
+        if (contactsIsAvailable()) {
+            contactsViewModel.onPermissionGranted();
+            contactsViewModel.loadAll();
+        }
+    }
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        if (!contactsIsAvailable()) {
```

Comment:
Also, this results in bad UX if "deny" is tapped.

Reply:
Well, it's indeed necessary for the fragment, but asking for permission repeatedly when the user says no is not something I ever saw in an app. I would probably delete such an app, too... Wouldn't you? :)

I suppose in general, an entity that wants to show something that may not be available should be able to handle such a state. For example, "can't access contacts" view. Or even an empty state that also covers the edge case of 0 contacts. 

For this specific project it doesn't matter yet, and we will probably see what they want to present in their designs, but this is something to consider.

Reply:
special empty view sounds good

Reply:
Another point regarding this - we will need a loading indicator for the duration of the fetching process. 

Also - we need to make sure the work of the contact provider isn't happening on the UI thread